nntrainer/tensor/manager.cpp

```diff
@@ -876,57 +799,64 @@ std::vector<Var_Grad *>
+    auto const &dim = inputs_dim[idx];
+    Tensor *var = nullptr, *grad = nullptr;
+    if (!outputs_name.empty()) {
+      var = tensor_pool.requestPrerequestedTensor(
+        dim, /// tensor dim
+        var_exec_order, var_ls,
+        outputs_name[idx],        /// name
+        Tensor::Initializer::NONE /// tensor initializer
+      );
 
-    std::transform(
-      inputs_dim.begin(), inputs_dim.end(), std::back_inserter(inputs_spec),
-      [&count, &node, &tspan](auto const &elem) {
-        return std::make_tuple(elem, Tensor::Initializer::NONE, true,
-                               node.getName() + std::string(":input") +
-                                 std::to_string(count++),
-                               tspan);
-      });
+      grad = tensor_pool.requestPrerequestedTensor(
+        dim, /// tensor dim
+        grad_exec_order, grad_ls,
+        outputs_name[idx] + Var_Grad::grad_suffix, /// name
+        Tensor::Initializer::ZEROS                 /// tensor initializer
+      );
+    } else if (!node.getInputConnections().empty()) {
+      /** skip requesting tensor for input */
+      const std::string &var_name =
+        node.getName() + std::string(":input") + std::to_string(idx);
+      var = tensor_pool.requestTensor(
+        dim, /// tensor dim
+        var_exec_order, var_ls,
+        var_name,                 /// name
+        Tensor::Initializer::NONE /// tensor initializer
+      );
 
-    ret = requestTensors<Var_Grad>(node, inputs_spec, inputs_v2);
-  } else {
-    ret.reserve(inputs_dim.size());
-
-    /**
-     * Find already allocated output which must match the name and dimensions
-     */
-    for (unsigned int idx = 0; idx < inputs_dim.size(); idx++) {
-      auto output_loc = name_map.at(outputs_name.at(idx));
-      Var_Grad *vg = outputs_v2.at(output_loc).get();
-      if (vg->getDim() != inputs_dim[idx])
-        throw std::invalid_argument(
-          "Dimension mismatch for the requested input");
-      ret.push_back(vg);
+      grad = tensor_pool.requestTensor(
+        dim, /// tensor dim
+        grad_exec_order, grad_ls,
+        var_name + Var_Grad::grad_suffix, /// name
+        Tensor::Initializer::ZEROS        /// tensor initializer
+      );
     }
-  }
-
-  const auto &exec_order = node.getExecutionOrder();
-  for (auto const &in : ret) {
-    auto const &vname = in->getName();
-    auto const &gname = in->getGradientName();
-
-    /** usage for inputs */
-    tensor_exec_order[vname].push_back(std::get<0>(exec_order));
-    tensor_exec_order[vname].push_back(std::get<1>(exec_order));
 
-    /** usage for inputs gradients (outgoing derivatives) */
-    tensor_exec_order[gname].push_back(std::get<2>(exec_order));
-
-    /** set tensor lifespan */
-    expandLifespan(vname, tspan);
-    expandLifespan(gname, tspan);
+    inputs_v2.emplace_back(std::make_unique<Var_Grad>(var, grad));
   }
 
+  std::transform(inputs_v2.begin() + current_size, inputs_v2.end(),
```

Comment:
minor optimization: `ret.reserve(current_size)`

Reply:
updated to `ret.reserve(inputs_dim.size())`.